In [3]:
%matplotlib notebook
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [4]:
t1a, t1b, t2a, t2b = -1, 3, -1.5, 1.5

t1s = np.linspace(t1a, t1b, 500)
t2s = np.linspace(t2a, t2b, 500)

# t1, t2 are both (500, 500) array
t1, t2 = np.meshgrid(t1s, t2s)

# it is (250000, 2) array
# from bottom left, to top right
T = np.c_[t1.ravel(), t2.ravel()]

Xr = np.array([[-1, 1], [-0.3, -1], [1, 0.1]])
Xr

yr = 2 * Xr[:, :1] + 0.5 * Xr[:, 1:]
yr

array([[-1.5 ],
       [-1.1 ],
       [ 2.05]])

$X_r$ are the samples, $y_r$ are the outputs. The model is $y_r = X_r \cdot (2, 0.5)^T$

$$X_r = 
\begin{pmatrix}
-1 & 1 \\
-0.3 & -1 \\
1 & 0.1 \\
\end{pmatrix} \\
y_r = \begin{pmatrix}
-1 & 1 \\
-0.3 & -1 \\
1 & 0.1 \\
\end{pmatrix}
\cdot
\begin{pmatrix}
2 \\
0.5
\end{pmatrix}
=
\begin{pmatrix}
-1.5 \\
-1.1 \\
-2.05
\end{pmatrix}
$$


### Now define the cost function, just MSE

* It's the MSE on each mesh grid.

In [7]:
(T.dot(Xr.T) - yr.T)

array([[ 1.        ,  2.9       , -3.2       ],
       [ 0.99198397,  2.89759519, -3.19198397],
       [ 0.98396794,  2.89519038, -3.18396794],
       ...,
       [ 0.01603206, -1.29519038,  1.08396794],
       [ 0.00801603, -1.29759519,  1.09198397],
       [ 0.        , -1.3       ,  1.1       ]])

In [8]:
(T.dot(Xr.T) - yr.T)**2

array([[1.00000000e+00, 8.41000000e+00, 1.02400000e+01],
       [9.84032193e-01, 8.39605789e+00, 1.01887617e+01],
       [9.68192899e-01, 8.38212734e+00, 1.01376518e+01],
       ...,
       [2.57027080e-04, 1.67751812e+00, 1.17498649e+00],
       [6.42567701e-05, 1.68375328e+00, 1.19242899e+00],
       [0.00000000e+00, 1.69000000e+00, 1.21000000e+00]])

In [10]:
np.sum((T.dot(Xr.T) - yr.T)**2, axis=1)

array([19.65      , 19.56885173, 19.48797206, ...,  2.85276164,
        2.87624652,  2.9       ])

In [13]:
J = (1. / len(Xr) * np.sum((T.dot(Xr.T) - yr.T)**2, axis=1)).reshape(t1.shape)
J

array([[6.55      , 6.52295058, 6.49599069, ..., 4.14168207, 4.1591296 ,
        4.17666667],
       [6.54112642, 6.51405772, 6.48707855, ..., 4.1232278 , 4.14065606,
        4.15817385],
       [6.53230128, 6.5052133 , 6.47821485, ..., 4.10482198, 4.12223095,
        4.13972947],
       ...,
       [8.10955578, 8.07292568, 8.0363851 , ..., 0.9396377 , 0.94750455,
        0.95546093],
       [8.12475368, 8.08810429, 8.05154444, ..., 0.94525491, 0.95310248,
        0.96103958],
       [8.14      , 8.10333134, 8.06675221, ..., 0.95092055, 0.95874884,
        0.96666667]])

### Now we get the MSE cost function @ each meshgrid

* And it's shape is the mesh shape